File: web-client/src/recommendations/augmentCache.ts
import { apiClient } from "../api/client";
import type { AugmentResponse, RecipeResult } from "../api/types";

// Per-session cache of recipe-direction enhancements, keyed by recipe id.
// We prefetch these as soon as recommendations arrive, so opening a recipe
// shows the enhanced steps instantly (no loading state).
const cache = new Map<number, Promise<AugmentResponse>>();

export function prefetchAugment(recipe: RecipeResult): Promise<AugmentResponse> {
  const existing = cache.get(recipe.id);
  if (existing) {
    return existing;
  }

  const promise = apiClient
    .augmentRecipe({
      title: recipe.title,
      ingredients: [...(recipe.inventory_matched ?? []), ...(recipe.inventory_missing ?? [])],
      directions: recipe.directions ?? [],
    })
    .catch((error) => {
      // Drop failures so a later open can retry.
      cache.delete(recipe.id);
      throw error;
    });

  cache.set(recipe.id, promise);
  return promise;
}

export function getAugment(recipeId: number): Promise<AugmentResponse> | undefined {
  return cache.get(recipeId);
}

export function prefetchAugmentAll(recipes: RecipeResult[]): void {
  for (const recipe of recipes) {
    // Swallow here; consumers attach their own handlers.
    void prefetchAugment(recipe).catch(() => {});
  }
}

export function clearAugmentCache(): void {
  cache.clear();
}

File: ios/Dishify/Features/Recommend/AugmentCache.swift
import Foundation

enum AugmentCache {
    private static var cache: [Int: Task<AugmentResponse, Error>] = [:]

    static func prefetchAll(_ recipes: [RecipeResult], api: APIClient = APIClient()) {
        for recipe in recipes {
            prefetch(recipe, api: api)
        }
    }

    @discardableResult
    static func prefetch(_ recipe: RecipeResult, api: APIClient = APIClient()) -> Task<AugmentResponse, Error>? {
        if let existing = cache[recipe.id] {
            return existing
        }

        let request = augmentRequest(for: recipe)
        let task = Task {
            try await api.augmentRecipe(request)
        }
        cache[recipe.id] = task
        return task
    }

    static func get(recipeId: Int) -> Task<AugmentResponse, Error>? {
        cache[recipeId]
    }

    static func remove(recipeId: Int) {
        cache.removeValue(forKey: recipeId)
    }

    static func clear() {
        cache.removeAll()
    }

    private static func augmentRequest(for recipe: RecipeResult) -> AugmentRequest {
        AugmentRequest(
            title: recipe.title,
            ingredients: (recipe.inventoryMatched ?? []) + (recipe.inventoryMissing ?? []),
            directions: recipe.directions ?? []
        )
    }
}
